### PR TITLE
Anf 12288 [NetAppFiles] Add preview api version 2021-04-01-preview

### PR DESCRIPTION
**Pull request by audunn**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by mentat9** [comment]
>         "storageToNetworkProximity": {

I'd suggest either unifying the two different "storageToNetworrkProximity" schema, or giving them different names. Since one allows mode 3 and one doesn't, they are spec'd to be different, but they have the same name so it's confusing for anyone using the API.

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2876 in 9ecd9cf. [](commit_id = 9ecd9cf8905d9451d44607340aa8d492cc2bf04e, deletion_comment = False)

**Comment by mentat9** [comment]
>     "regionInfo": {

I'm confused about what happened here. According to the PR diffs, this is a new definition that didn't exist in previous API versions. However, there is a $ref to this that apparently existed previously. How did the path that references this definition get added without this definition?

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2872 in 9ecd9cf. [](commit_id = 9ecd9cf8905d9451d44607340aa8d492cc2bf04e, deletion_comment = False)

**Comment by mentat9** [comment]
>           "type": "integer",

Highly recommend making this a string enum and naming the values it can take. Mode 0 doesn't convey the meaning of the value. With modelAsString, these will have string values "0", "1", etc. which is also weird.

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2879 in 9ecd9cf. [](commit_id = 9ecd9cf8905d9451d44607340aa8d492cc2bf04e, deletion_comment = False)

**Comment by mentat9** [comment]
>           "type": "integer",

Same comment here regarding named string enum vs. numeric.

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:3746 in 9ecd9cf. [](commit_id = 9ecd9cf8905d9451d44607340aa8d492cc2bf04e, deletion_comment = False)

**Comment by mentat9** [comment]
>         "storageToNetworkProximity": {

I now see the later iteration (after the file was renamed) where you did make these different names.

---
In reply to: [833832648](https://github.com/Azure/azure-rest-api-specs/pull/14286#issuecomment-833832648) [](http://example.com/codeflow?ancestors=833832648)

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2876 in b778c91. [](commit_id = b778c919cc26bda039a85503bdf97a0c8eb41912, deletion_comment = False)

**Comment by audunn** [comment]
> I'm confused about what happened here. According to the PR diffs, this is a new definition that didn't exist in previous API versions. However, there is a $ref to this that apparently existed previously. How did the path that references this definition get added without this definition?

That is interesting where do you see this?

**Comment by audunn** [comment]
> > ```
> >       "type": "integer",
> > ```
> 
> Highly recommend making this a string enum and naming the values it can take. Mode 0 doesn't convey the meaning of the value. With modelAsString, these will have string values "0", "1", etc. which is also weird.
> 
> Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2879 in [9ecd9cf](https://github.com/Azure/azure-rest-api-specs/commit/9ecd9cf8905d9451d44607340aa8d492cc2bf04e). [](commit_id = [9ecd9cf](https://github.com/Azure/azure-rest-api-specs/commit/9ecd9cf8905d9451d44607340aa8d492cc2bf04e), deletion_comment = False)

I recommended the same to the internal team implementing this, according to them this was on request from MSFT

**Comment by mentat9** [comment]
>             "name": "RegionStorageToNetworkProximity",

I think this and the other rename of the x-ms-enum below may break SDKs, so that might come up downstream. Not an issue for ARM review, just a heads-up.

---
Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01-preview/netapp.json:2888 in 2149d85. [](commit_id = 2149d85c047fa4b1933f5430bc41efb709f52189, deletion_comment = False)

**Comment by mentat9** [comment]
> > I'm confused about what happened here. According to the PR diffs, this is a new definition that didn't exist in previous API versions. However, there is a $ref to this that apparently existed previously. How did the path that references this definition get added without this definition?
> 
> That is interesting where do you see this?

By diffing iteration 1 and 7, it shows only changes added (before the file was moved/renamed). In that diff, the GET definition for this path
```
"/subscriptions/{subscriptionId}/providers/Microsoft.NetApp/locations/{location}/regionInfo"
```
shows as unchanged (present in initial check-in), but references the "RegionInfo" schema, which shows as new. It doesn't matter too much, because the latest iteration has both. Just think something was amiss with the first commit.

**Comment by mentat9** [comment]
> > > ```
> > >       "type": "integer",
> > > ```
> > 
> > 
> > Highly recommend making this a string enum and naming the values it can take. Mode 0 doesn't convey the meaning of the value. With modelAsString, these will have string values "0", "1", etc. which is also weird.
> > Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2879 in [9ecd9cf](https://github.com/Azure/azure-rest-api-specs/commit/9ecd9cf8905d9451d44607340aa8d492cc2bf04e). [](commit_id = [9ecd9cf](https://github.com/Azure/azure-rest-api-specs/commit/9ecd9cf8905d9451d44607340aa8d492cc2bf04e), deletion_comment = False)
> 
> I recommended the same to the internal team implementing this, according to them this was on request from MSFT

That's unfortunate, but I won't block approval on it. By MSFT, you mean someone from the NetApp team? Can you email me a contact?

File: specification/netapp/resource-manager/readme.go.md [hunk 1]
```diff
@@ -30,6 +31,15 @@ batch:
   - tag: package-netapp-2019-05-01
   - tag: package-2017-08-15
 ```
+### Tag: package-netapp-2021-04-01-preview and go
+
+These settings apply only when `--tag=package-netapp-2021-04-01-preview --go` is specified on the command line.
+Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+``` yaml $(tag) == 'package-netapp-2021-04-01-preview' && $(go)
+output-folder: $(go-sdk-folder)/services/$(namespace)/mgmt/2021-04-01-preview/$(namespace)
```
Comment (ArcturusZhang) [inline specification/netapp/resource-manager/readme.go.md:40]: We need to have the preview packages in the preview sub-directory
```suggestion
output-folder: $(go-sdk-folder)/services/preview/$(namespace)/mgmt/2021-04-01-preview/$(namespace)
```

**Comment by audunn** [comment]
> I think this and the other rename of the x-ms-enum below may break SDKs, so that might come up downstream. Not an issue for ARM review, just a heads-up.

There are no SDK's yet for this is all new in this preview version.

**Comment by audunn** [comment]
@mentat9 @ArcturusZhang  Note I updated the spec to reflect the recently merged 2021-02-01 public api-changes.

I am seeking more information on storageToNetworkProximity  enum questions.

**Comment by soorajchirag** [comment]
> > > > ```
> > > >       "type": "integer",
> > > > ```
> > > 
> > > 
> > > Highly recommend making this a string enum and naming the values it can take. Mode 0 doesn't convey the meaning of the value. With modelAsString, these will have string values "0", "1", etc. which is also weird.
> > > Refers to: specification/netapp/resource-manager/Microsoft.NetApp/preview/2021-04-01/netapp.json:2879 in [9ecd9cf](https://github.com/Azure/azure-rest-api-specs/commit/9ecd9cf8905d9451d44607340aa8d492cc2bf04e). [](commit_id = [9ecd9cf](https://github.com/Azure/azure-rest-api-specs/commit/9ecd9cf8905d9451d44607340aa8d492cc2bf04e), deletion_comment = False)
> > 
> > 
> > I recommended the same to the internal team implementing this, according to them this was on request from MSFT
> 
> That's unfortunate, but I won't block approval on it. By MSFT, you mean someone from the NetApp team? Can you email me a contact?

@mentat9  The contacts for this in MSFT would be Ram Kakani and Shane Manali.

**Comment by ChenTanyi** [comment]
@audunn It seems you had added a new API (2021-02-01) after ARM sign off. It may need to ask ARM to sign off again. Thanks!

**Comment by audunn** [comment]
@ChenTanyi That came as we merge/synced from master that api-version was added in another pr. https://github.com/Azure/azure-rest-api-specs/pull/13814

**Comment by ChenTanyi** [comment]
> @ChenTanyi That came as we merge/synced from master that api-version was added in another pr. #13814

Got it, I would approve this time. In the next time, if you didn't change anything, you **should not** add it into this PR.

**Review by ChenTanyi — APPROVED** [review]

**Comment by ChenTanyi** [comment]
@audunn please let me know when it ready to merge.

**Comment by audunn** [comment]
@ChenTanyi  I just pushed a change to address a comment from the ARM Api reviewer on `storageToNetworkProximity` to align with their comments.  Should be ready now.

**Comment by audunn** [comment]
@ChenTanyi  good to go, thank you.

**Pull request merged**